### PR TITLE
[ASTContext] Add setter for PreModuleImportCallback

### PR DESCRIPTION
**Pull request by chelcassanova**
`PreModuleImportCallback` is a variable for a callback function called before module importing in the Swift compiler. It is currently being used by LLDB to report progress updates on importing Swift modules in LLDB. This commit adds a setter for `PreModuleImportCallback` for greater flexibility in LLDB and removes the reference to this variable in the constructor.

Related to https://github.com/apple/llvm-project/pull/7769

rdar://105286354

### CODE REVIEW

File: include/swift/AST/ASTContext.h [hunk 1]
```diff
@@ -239,8 +239,8 @@ class ASTContext final {
       ClangImporterOptions &ClangImporterOpts,
       symbolgraphgen::SymbolGraphOptions &SymbolGraphOpts,
       SourceManager &SourceMgr, DiagnosticEngine &Diags,
-      llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> OutBackend = nullptr,
-      std::function<bool(llvm::StringRef, bool)> PreModuleImportCallback = {});
+      llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> OutBackend = nullptr
+      );
```
Comment (medismailben) [inline include/swift/AST/ASTContext.h:243]: Interesting formatting. Is that really `clang-format`'s output ?

File: lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -715,6 +714,10 @@ ASTContext::~ASTContext() {
   getImpl().~Implementation();
 }
 
+void ASTContext::SetPreModuleImportCallback(std::function<bool(llvm::StringRef ModuleName, bool IsOverlay)> callback) {
+  ASTContext::PreModuleImportCallback = callback;
```
Comment (medismailben) [inline lib/AST/ASTContext.cpp:718]: Why do you need to qualify `PreModuleImportCallback` here ?

**Comment by medismailben** [comment]
FWIW, although this callback is used mainly as an LLDB module loading progress callback, it could also be used for/by other things. I'd retitle this PR to be more generic.

File: include/swift/AST/ASTContext.h (hunk 1 above)
Comment (chelcassanova) [reply include/swift/AST/ASTContext.h:243]: Yeah it is, other than removing the callback in the constructor I didn't change anything else here.

File: lib/AST/ASTContext.cpp (hunk 2 above)
Comment (chelcassanova) [reply lib/AST/ASTContext.cpp:718]: I don't, but I think I originally thought that `PreModuleImportCallback` was static even though it's not so I can remove the qualification.

**Comment by chelcassanova** [comment]
@swift-ci test

**Comment by chelcassanova** [comment]
Good point about `PreModuleImportCallback` being generic. Since LLDB was the only one use it I thought it was created specifically for LLDB, I'll update the PR description to reflect this.

File: include/swift/AST/ASTContext.h [hunk 3]
```diff
@@ -294,6 +294,8 @@ class ASTContext final {
   /// OutputBackend for writing outputs.
   llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> OutputBackend;
 
+  void SetPreModuleImportCallback(std::function<bool(llvm::StringRef ModuleName, bool IsOverlay)> callback);
```
Comment (adrian-prantl) [inline include/swift/AST/ASTContext.h:297]: Doxygen comment?

**Review by adrian-prantl — APPROVED** [review]

File: include/swift/AST/ASTContext.h (hunk 3 above)
Comment (chelcassanova) [reply include/swift/AST/ASTContext.h:297]: Yes, I'll add one here

**Comment by chelcassanova** [comment]
@swift-ci test

**Comment by chelcassanova** [comment]
@swift-ci smoke test

**Comment by chelcassanova** [comment]
https://github.com/apple/llvm-project/pull/7769
@swift-ci  test

**Comment by chelcassanova** [comment]
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

**Comment by chelcassanova** [comment]
https://github.com/apple/llvm-project/pull/7769
@swift-ci please smoke test

File: lib/AST/ASTContext.cpp [hunk 4]
```diff
@@ -715,6 +714,10 @@ ASTContext::~ASTContext() {
   getImpl().~Implementation();
 }
 
+void ASTContext::SetPreModuleImportCallback(std::function<void(llvm::StringRef ModuleName, bool IsOverlay)> callback) {
```
Comment (augusto2112) [inline lib/AST/ASTContext.cpp:717]: Was this line `clang-format`ted? Looks pretty long

Comment (chelcassanova) [reply lib/AST/ASTContext.cpp:717]: It should've been, but I can also shorten it myself if `clang-format` is ignoring it for some reason

Comment (augusto2112) [reply lib/AST/ASTContext.cpp:717]: If `clang-format` says it's ok then it's fine

**Comment by chelcassanova** [comment]
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

**Comment by chelcassanova** [comment]
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

**Comment by chelcassanova** [comment]
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

**Comment by chelcassanova** [comment]
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

**Comment by chelcassanova** [comment]
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

**Pull request merged**